mediatek/filogic: add support for Motorcomm PHYs

aimacintyre opened the pull request:
There appears to have been a silent change (i.e. no version number change) to some Cudy Filogic based devices (at least WR3000H v1 and M3000 v2 have been reported) replacing RTL8221B PHYs with Motorcomm YT8821 PHYs for the 2.5GBps capable WAN ports and without the Motorcomm PHY driver have no apparent WAN port.  The current DTS sets the WAN port as a generic clause 45 PHY so enable inclusion of the Motorcomm PHY driver; this may be sufficient to get a working LAN port on it's own.  There appears to have been no recent OEM firmware release for potentially affected models so as yet there is no way to tell if DTS changes are required.

If accepted this should be backported to the 25.12 branch as the 6.12 kernel has YT8821 driver support.

Fixes: https://github.com/openwrt/openwrt/issues/21244
Fixes: https://github.com/openwrt/openwrt/issues/21051
       (M3000v2 not officially supported yet but M3000v1 24.10
        images work satisfactorily on M3000v2 devices with
        RTL8221B PHY and available Cudy OEM firmware releases don't
        distinguish between v1 and v2)

DragonBluep commented:
This won't work. I've already tried this change. Report: https://github.com/openwrt/openwrt/issues/21244#issuecomment-3685748889

aimacintyre committed:
kernel: add kmod-phy-motorcomm

The Motorcomm YT8821 2.5G PHY is being used in some devices as an
alternative to devices like the RTL8221B.  Package it as a kmod
so it can be used as a device package rather than requiring changes
to the target kernel config.

Signed-off-by: Andrew MacIntyre <andymac@pcug.org.au>

aimacintyre committed:
mediatek/filogic: Motorcomm PHY support for Cudy WR3000H v1

Recent Cudy WR3000H v1 units have a Motorcomm YT8821 2.5G PHY for
the WAN interface instead of the RTL8221B PHY used in earlier units,
so add kmod-phy-motorcomm to be able to use it.

Signed-off-by: Andrew MacIntyre <andymac@pcug.org.au>

aimacintyre commented:
> Is the MDIO address still the same?

Yes

aimacintyre commented:
> On the Cudy WR300H v1, this PR did not produce any results.

Could you please attach the bootlog from the build you tested.

aimacintyre commented:
> ```
> [3 янв. 2026 г., 00:12:37 UTC] kern.info: [   12.680688] mtk_soc_eth 15100000.ethernet wan: PHY [mdio-bus:06] driver [Generic Clause 45 PHY] (irq=POLL)
> [3 янв. 2026 г., 00:12:37 UTC] kern.info: [   12.690403] mtk_soc_eth 15100000.ethernet wan: configuring for phy/2500base-x link mode
> [3 янв. 2026 г., 00:12:37 UTC] daemon.notice: netifd: Interface 'wan' is enabled
> [3 янв. 2026 г., 00:12:37 UTC] daemon.notice: netifd: Interface 'wan6' is enabled
> ```

It hasn't attached the Motorcomm PHY driver, falling back on generic code which isn't sufficient to actually set things up properly, not least that I suspect it should have an IRQ assigned (as is the case for the RTL8221B PHY) rather than be set for POLLing.  Closing this as I can't take this any further.

DJs3000 commented:
@aimacintyre Andrew, your PR needs to be completed with just one line. 
https://github.com/openwrt/openwrt/blob/ded99c2984b245ee3885bdf10bd1797e838abf54/target/linux/mediatek/dts/mt7981b-cudy-wr3000h-v1.dts#L106
replace to:
compatible = "ethernet-phy-id4f51.ea19";

JakubVanek commented:
I found some stuff that indicates that previous revisions of WR3000H used a Realtek PHY. In the (now somewhat outdated) Cudy WR3000H GPL tarball, the device tree file corresponding to WR3000H seems to be `R63.dts` and the `compatible` string there is only `compatible = "ethernet-phy-id001c.c849";`. This ID matches the [RTL8221B-VB-CG ID](https://github.com/torvalds/linux/blob/aacb0a6d604ac9953b261963efe25ae7521b7c58/drivers/net/phy/realtek/realtek_main.c#L191).

DJs3000 commented:
@JakubVanek I can't install mdio-tools
<img width="602" height="333" alt="image" src="https://github.com/user-attachments/assets/1fef8477-da7b-403b-af70-23458ef7a3f4" />

DJs3000 commented:
@JakubVanek I will now rebuild owrt with "c22" and return the result.

JakubVanek commented:
Thank you, this seems to rule out the hypothesis about the ID registers -  the IDs on both c22 and c45 are correct.. I am then not sure why the PHY is not getting auto-detected correctly.

JakubVanek commented:
There seems to be one more workaround - moving the reset line definition from the PHY device to the MDIO bus: https://lore.kernel.org/all/20251119134750.394655-1-horatiu.vultur@microchip.com/
